### PR TITLE
Island building from scratch

### PR DESCRIPTION
**Pull request by geralddejong**
Huge changes, but now there's a nice state machine.  You can now empty local storage and start a new island.  Page refresh to try another random attempt, fix the one you think is ok. Claim it. Then "logout" and claim others.

### CODE REVIEW

File: client/src/app-storage.ts [hunk 1]
```diff
@@ -4,8 +4,6 @@ import {Hexalot} from "./island/hexalot"
 import {Island, IslandPattern} from "./island/island"
 import {Journey} from "./island/journey"
 
-const MASTER_KEY = "master"
```
Comment (mitchelljustin) [inline client/src/app-storage.ts:7]: What replaces the concept of "Master"?

File: client/src/app-storage.ts [hunk 2]
```diff
@@ -69,7 +58,7 @@ export class AppStorage {
         }
     }
 
-    public loadJourney(hexalot: Hexalot, island: Island): void {
+    public loadJourney(hexalot: Hexalot, island: Island): Journey | undefined {
```
Comment (mitchelljustin) [inline client/src/app-storage.ts:61]: Might just be my preference, but null is a saner sentinel value.
```suggestion
    public loadJourney(hexalot: Hexalot, island: Island): Journey | null {
```

File: client/src/app-storage.ts [hunk 3]
```diff
@@ -23,15 +21,6 @@ export class AppStorage {
...
-    public getMaster(): string | undefined {
-        const value = this.storage.getItem(MASTER_KEY)
-        return value ? value : undefined
-    }
-
-    public setMaster(master: string): void {
-        this.storage.setItem(MASTER_KEY, master)
-    }
-
     public getPhysicsFeature(feature: PhysicsFeature): number {
```
Comment (mitchelljustin) [inline client/src/app-storage.ts:24]: Yer gonna have to make these storage methods async if you wanna load them from the server. Will that be easy right now?

File: client/src/app-storage.ts (hunk 1 above)
Comment (geralddejong) [reply client/src/app-storage.ts:7]: Nothing for now.  I can haz Twitter handle?

File: client/src/app-storage.ts (hunk 2 above)
Comment (geralddejong) [reply client/src/app-storage.ts:61]: I don't see a justification.  The optional? syntax uses undefined, so I go with it.

File: client/src/app-storage.ts (hunk 3 above)
Comment (geralddejong) [reply client/src/app-storage.ts:24]: Done that a few times.  Not hard at all.  I use lots of asyc now even including loading the WASM module.

File: client/src/gotchi/evolver.ts [hunk 4]
```diff
@@ -67,7 +67,7 @@ export class Evolver {
     }
 
     public mutateGenome(mutationCount: number): void {
-        console.log(`mutating ${this.index} ${Direction[this.nextDirection]} ${mutationCount} dice`)
+        // console.log(`mutating ${this.index} ${Direction[this.nextDirection]} ${mutationCount} dice`)
```
Comment (mitchelljustin) [inline client/src/gotchi/evolver.ts:70]: Remove

File: client/src/index.css [hunk 5]
```diff
@@ -93,18 +99,28 @@ p a, p a:hover {
...
 .info-panel {
+    background-color: black;
     bottom: 20px;
-    left: 20px;
-    width: 30%;
-    height: 35%;
+    left: 10%;
+    width: 80%;
+    height: auto;
+    padding: 5px 5px 5px 5px;
```
Comment (mitchelljustin) [inline client/src/index.css]:
```suggestion
    padding: 5px;
```

File: client/src/index.css [hunk 6]
```diff
@@ -93,18 +99,28 @@ p a, p a:hover {
...
+    border-color: darkcyan;
+    border-style: solid;
+    border-radius: 10px;
 }
 
 .info-panel-collapsed {
     bottom: 20px;
     left: 20px;
+    padding: 5px 5px 5px 5px;
```
Comment (mitchelljustin) [inline client/src/index.css]:
```suggestion
    padding: 5px;
```

File: client/src/gotchi/evolver.ts (hunk 4 above)
Comment (geralddejong) [reply client/src/gotchi/evolver.ts:70]: I'd rather keep it, so i can activate while debugging.

File: client/src/island/hexalot.ts [hunk 7]
```diff
@@ -83,34 +83,30 @@ export class Hexalot {
...
-        }
-        return this.genomeStored
+    public get occupied(): boolean {
+        return !!this.genome
     }
 
-    public deleteGenome(): void {
-        this.genomeStored = undefined
+    public get isLegal(): boolean {
+        return !this.spots.some(spot => !spot.isLegal)
```
Comment (mitchelljustin) [inline client/src/island/hexalot.ts]: Too many negatives

```suggestion
        return this.spots.every(spot => spot.isLegal)
```

File: client/src/app-storage.ts (hunk 2 above)
Comment (mitchelljustin) [reply client/src/app-storage.ts:61]: I don't really care which of `null` or `undefined` is used, i'm just annoyed that we have both. Let's enable this lint rule? https://palantir.github.io/tslint/rules/no-null-keyword/

File: client/src/gotchi/evolver.ts (hunk 4 above)
Comment (mitchelljustin) [reply client/src/gotchi/evolver.ts:70]: Mkay

File: client/src/view/gotchi-view.tsx [hunk 8]
```diff
@@ -115,90 +108,38 @@ export class GotchiView extends React.Component<IGotchiViewProps, IGotchiViewSta
...
-    //     };
-    //     const geometry = new Geometry();
-    //     const spot = this.props.selectedSpot.getValue();
-    //     const action = this.props.evolution || this.props.gotchi;
-    //     if (spot && this.orbit && !this.orbit.changing && !action) {
-    //         const target = spot.centerOfHexalot ? new Vector3(0, HUNG_ALTITUDE, 0).add(spot.center) : spot.center;
-    //         geometry.vertices = [userCoords(0), target, userCoords(-1), target, userCoords(1), target];
-    //     }
-    //     return geometry;
-    // }
```
Comment (mitchelljustin) [inline client/src/view/gotchi-view.tsx:178]: Remove

File: client/src/island/spot.ts [hunk 9]
```diff
@@ -165,6 +111,90 @@ export class Spot {
...
+    private get normalSpread(): number {
+        switch (this.surface) {
+            case Surface.Land:
+                return LAND_NORMAL_SPREAD
+            case Surface.Water:
+                return WATER_NORMAL_SPREAD
+            default:
+                return 0
+        }
+    }
```
Comment (mitchelljustin) [inline client/src/island/spot.ts:197]: This is all voodoo code to me, haha

File: client/src/island/island.ts [hunk 10]
```diff
@@ -13,41 +14,25 @@ export interface IslandPattern {
...
     }
 
-    public get hexalotsWithSeeds(): Hexalot[] {
-        return this.hexalots.filter((hexalot: Hexalot, index: number) => !!hexalot.master)
-    }
-
-    public get isLegal(): boolean {
-        return !this.spots.find(spot => !spot.legal)
+    public get islandIsLegal(): boolean {
+        return !this.spots.some(spot => !spot.isLegal)
```
Comment (mitchelljustin) [inline client/src/island/island.ts]:
```suggestion
        return this.spots.every(spot => spot.isLegal)
```

**Review by mitchelljustin — APPROVED** [review]

File: client/src/app-storage.ts (hunk 2 above)
Comment (geralddejong) [reply client/src/app-storage.ts:61]: yes!

**Pull request merged**